### PR TITLE
Add phone-number library

### PR DESCRIPTION
**Pull request by umurkontaci**
This PR is part of #7453 and adds a phone number library for parsing and formatting numbers.

I was able to simplify this to a point where I am almost satisfied. There will be tons of things to improve I believe, but I think it is vastly better than what's out in the wild.

This one comes with a vast amount of research done on other libraries to understand the phone numbers field.

Localized phone numbers are a beast and there are some libraries out there that does the same thing, but either they are terribly written, or they are huge behemoths that I don't want to include them in our builds. The most prominent one is [libphonenumber](https://github.com/googlei18n/libphonenumber), while it is super comprehensive, it is Java-centric and lacks quite a bit documentation, uses Clojure compiler, and has a size of *350 KB* when minified and *70KB* when gzipped.

This library is stateless and I think it has a clear API with a sound structure.
 
In `bin/build-metadata.js` we are consuming libphonenumber's metadata files and creating our own stripped down version. The file looks very big but it is very redundant hence very compressible (`8KB` minified and gzipped).

There might still be quirks around edge cases during formatting, so please do try to break it :)

Anything that breaks should be added as a test, fixed and verified to prevent regressions in the future.

/cc: @klimeryk @aidvu @wensco @Automattic/sdev-feed 

Test live: https://calypso.live/?branch=add/phone-number-library

### CODE REVIEW

**Comment by yoavf** [comment]
Drive by comment, just in case you weren't aware: two factor authentication currently use https://github.com/Automattic/node-phone, a fork of https://github.com/AfterShip/phone

**Comment by klimeryk** [comment]
Thanks for the heads-up, @yoavf. I'm not familiar with any of them (yet!), but I'd love your opinion and assesement of the solution that Umur prepared here. Pinging also @southp, @jmdodd, @dllh, @ebinnion and @allendav who contributed to our fork of `AfterShip/phone`. Please have a look at this PR (and #7456 that builds upon it) to see if it could be used on the 2FA page too. It'd be best if we could agree on one library for parsing/validating phones as well as the component for inputting them.

**Comment by ebinnion** [comment]
Hi 👋 .

My main concern with using this over `node-phone` is completeness. Here's an example of all of the locales that `node-phone` handles: https://github.com/AfterShip/phone/blob/master/lib/index.js

And even after all of that, there are still some edge cases we have run into that forced us to fork `node-phone` so we could quickly update when we needed to. Although, that has also led to an issue where we forget to go back and update our fork. 😞 

Unless I misread, I don't think this PR handles all of the cases as `node-phone`. If that is the case, I think we'd want to reconsider though. 

Something that I've wanted to add for a while was a fallback to Twilio's lookup API. So, if a number fails, we then call the Twilio's API to validate the number for us (since they're the ones that actually send all of our SMS messages). Based on the pricing, format lookups are free too. So, this way we get the benefit of a super fast in-browser validation backed up by a more complete API-based validation.

To summarize, I think we should make sure that we're covering all of the cases as `node-phone`. And if not, let's either 1) cover those or 2) fallback to Twilio's lookup API.

**Comment by dllh** [comment]
I share Eric's concerns. This looks nice and light, but unless I'm misreading it, it supports only 3 countries at present, and we send 2fa requests to many countries. I fear that making a new lib would be pretty time intensive (not sure what your timeframe for this is) or would lead to regressions. The Twilio fallback could be a nice compromise.

**Comment by umurkontaci** [comment]
@yoavf @dllh Thanks for dropping by :) I don't think completeness is going to be an issue as this should support everything out there. If a country does not have a local format or we don't have the data, it is treated as-is (no formatting). 

You can see the list of countries it supports here: https://github.com/Automattic/wp-calypso/blob/1c4664f8b6b5661afc46dcf72b59e7ce5dafa38a/client/components/phone-input/data.js

As for validations on Twilio, I think that should be handled elsewhere, possibly by the form controller or equivalent. Partly because I think it is out of scope and partly because I'd like to avoid adding asynchronous functions here.

**Comment by southp** [comment]
Hey, thanks for the effort of finding this and sorry for the late :)

In general, I think finding a better supported library instead of keeping maintaining our own node-phone fork is a great idea. It's from Google and the activity rate looks good, so it could be a good candidate.

I have two concerns about it though:

### 1. How the code is integrated?
It looks like the code is built manually and put under `bin` and `components` manually. I am worrying that it would mean to repeat the build & replace manual process for each major release. Also, it might becoming unclear that it is from 3rd-party in the future. Could it be possible to integrate it just like other 3rd-party libraries? To be honest, I haven't played the library, so I have no idea if that is possible :)

### 2. Our resolution against false verification could be limited.
The benefit of maintaining our own verification fork is that we can react against false verification very fast. However, once we rely on `libphonenumber`, it could be prolonging. For example, I spot-checked [their issue tracker](https://github.com/googlei18n/libphonenumber/issues) and saw many verification issues. Notably, while they are failing to verify [this](https://github.com/googlei18n/libphonenumber/issues/1252) and [this](https://github.com/googlei18n/libphonenumber/issues/1276), we can verify them without problems. I can see that they resolves verification issues quite rigorously ( e.g. they even do contact officials. ), so in the long term I believe it would serve as a better library. However, given the above two failed cases, I am worrying that it could break some verifications and we can't resolve them fast enough.

Finally, I think it is still worth approaching, since #7453 seems to need its formatting functionality  and it is not something `node-phone` has. A quick way to work around the coverage issue might be fallback to `node-phone` on failures. Might be a stupid idea, though :)

**Comment by umurkontaci** [comment]
@southp Thanks for dropping by :)


> It looks like the code is built manually and put under bin and components manually. I am worrying that it would mean to repeat the build & replace manual process for each major release. Also, it might becoming unclear that it is from 3rd-party in the future. Could it be possible to integrate it just like other 3rd-party libraries? To be honest, I haven't played the library, so I have no idea if that is possible :)

The code itself (business logic) is not being built from `libphonenumber`, the metadata is. Basically the file here: https://github.com/googlei18n/libphonenumber/blob/master/javascript/i18n/phonenumbers/metadatalite.js , which is some kind of protobuf, is converted to a proper JS object that we can use and access normally. Protobuf changes / upgrades are supposed to be support older versions. Even though protobuf doesn't introduce breaking changes by design, we are definitely not using them as advertised it is possible that it may break during an update. The chances are low, but it is possible.

I'm not too worried about it though, in case a metadata build fails, we will have plenty of errors in the test suite and we can fix the bug or pin to a working commit on `libphonenumber` until we find a solution.


> Our resolution against false verification could be limited.

It is true we will rely on Google to fix the upstream before we can do anything. But it is a 6 year old library, so I expect them to have figured out most of the issues :)

That being said, I didn't intend this to be a validator, phone number validation is a huge beast when we consider all the possible variations that are supported. Every form is going to have a different type of restriction on validation. For example, 2FA requires the phone numbers to be mobile phone numbers and their rules don't vary much, but there are other types of numbers (internal dial codes in companies used as suffix etc. +1 425 394-3491-341) we need to support in other forms. 

I think the validation should be handled outside of this library, as it will be hard to satisfy all the requirements on every different form.

**Comment by umurkontaci** [comment]
@klimeryk @ebinnion @southp @dllh I'm going forward with this unless you have any objections.

@klimeryk I didn't change anything, but would like a final green light :)

**Comment by southp** [comment]
@umurkontaci Hey, sorry that I didn't get back sooner. It looks like you are going to use it to format, not to validate, which I was misunderstanding. I am totally okay with it as long as it is not for validation :)

**Comment by klimeryk** [comment]
LGTM! Very nice and comprehensive battery of tests 👍
I'll leave more detailed review of the code for the phone input field PR (maybe including some profiling, given the large amount of string operations and regexes). But to move this awesome effort forward - merge away!
